network/src/peer/fuzzing.rs

```diff
@@ -0,0 +1,150 @@
+    // runtime and sometimes blocks when trying to shutdown the runtime.
+    //
+    // https://github.com/tokio-rs/tokio/pull/2649
+    let mut rt = tokio::runtime::Builder::new()
+        .basic_scheduler()
+        .enable_all()
+        .build()
+        .unwrap();
+    let executor = rt.handle().clone();
+
+    // We want to choose a constant peer id for _our_ peer id, since we will
+    // generate unbounded metrics otherwise and OOM during fuzzing.
+    let peer_id = PeerId::ZERO;
+    // However, we want to choose a random _remote_ peer id to ensure we _don't_
+    // have metrics logging the remote peer id (which would eventually OOM in
+    // production for public-facing nodes).
+    let remote_peer_id = PeerId::random();
+
+    // Mock data
+    let network_context = NetworkContext::mock_with_peer_id(peer_id);
+    let socket = ReadOnlyTestSocketVec::new(data.to_vec());
+    let metadata = ConnectionMetadata::new(
+        remote_peer_id,
+        ConnectionId::from(123),
+        NetworkAddress::mock(),
+        ConnectionOrigin::Inbound,
+        MessagingProtocolVersion::V1,
+        SupportedProtocols::from(
+            [
+                ProtocolId::ConsensusRpc,
+                ProtocolId::ConsensusDirectSend,
+                ProtocolId::MempoolDirectSend,
+                ProtocolId::StateSynchronizerDirectSend,
+                ProtocolId::DiscoveryDirectSend,
+                ProtocolId::HealthCheckerRpc,
+            ]
+            .iter(),
+        ),
+    );
+    let connection = Connection { socket, metadata };
+
+    let (connection_notifs_tx, connection_notifs_rx) = channel::new_test(8);
+    let max_concurrent_reqs = 8;
+    let max_concurrent_notifs = 8;
+    let channel_size = 8;
+
+    // Spin up a new `Peer` actor through the `NetworkProvider` interface.
+    let (network_reqs_tx, network_notifs_rx) = NetworkProvider::start(
+        network_context,
+        executor,
+        connection,
+        connection_notifs_tx,
+        max_concurrent_reqs,
+        max_concurrent_notifs,
+        channel_size,
+        MAX_FRAME_SIZE,
+    );
+
+    rt.block_on(async move {
+        // Wait for "remote" to disconnect (we read all data and socket read
+        // returns EOF), we read a disconnect request, or we fail to deserialize
+        // something.
+        let connection_notifs = connection_notifs_rx.collect::<Vec<_>>().await;
+
+        // ACK the "remote" d/c and drop our handle to the Peer actor. Then wait
+        // for all network notifs to drain out and finish.
+        drop(network_reqs_tx);
+        let network_notifs = network_notifs_rx.collect::<Vec<_>>().await;
+    });
+
+    // All tasks should be complete. The 60s timer here will at least save this
+    // fuzzer run as a "slow" run if we leak a task. Unfortunately there doesn't
+    // seem to be a way to Err/panic if there are leaked tasks...
+    rt.shutdown_timeout(Duration::from_secs(60));
```

Comment:
this prevents us from detecting real timeouts : o what do you mean by leak a task?